test/ctia/task/migration/migrations/describe_test.clj

```diff
@@ -0,0 +1,203 @@
+   :labels ["tool"],
+   :schema_version "1.0.16",
+   :type "tool",
+   :x_mitre_aliases ["PoshC2"],
+   :source "MITRE Enterprise ATT&CK",
+   :external_ids
+   ["tool--4b57c098-f043-4da2-83ef-7588a6d426bc"
+    "hydrant-05856bc0326b15f3d99036be057f9f7d06794dc57cac2f8f9d21c365c03f4d44"
+    "ATT&CK-S0378"],
+   :external_references
+   [{:source_name "mitre-attack",
+     :url "https://attack.mitre.org/software/S0378",
+     :external_id "S0378"}
+    {:source_name "GitHub PoshC2",
+     :description
+     "Nettitude. (2018, July 23). Python Server for PoshC2. Retrieved April 23, 2019.",
+     :url "https://github.com/nettitude/PoshC2_Python"}],
+   :source_uri "https://attack.mitre.org",
+   :id
+   "https://intel.amp.cisco.com:443/ctia/tool/tool-0b32dbc5-7207-4ffd-a484-ec6465d62586",
+   :tlp "green",
+   :groups ["tenzin"],
+   :timestamp "2020-03-30T02:37:23.626Z",
+   :owner "Tenzin"})
+
+
+(defn check-unchanged-fields
+  [source migrated]
+  (is (= (dissoc source :name :title :description :short_description)
+         (dissoc migrated :name :title :description :short_description))
+      "field not related to entity description shall be preserved."))
+
+(defn mutate-fn
+  [mutated-fields]
+  (fn [entity]
+    (let [properties (shuffle mutated-fields)
+          nb-mutations (-> (count mutated-fields)
+                           inc
+                           rand-int)]
+      (->> (take nb-mutations properties)
+           (apply (partial dissoc entity))))))
+
+(defn mutant-generator
+  [mutated-fields]
+  (let [mutate (mutate-fn mutated-fields)]
+    (fn [n entity] (repeatedly n #(mutate entity)))))
+
+(deftest migrate-non-describable-test
+  (testing "describe shall properly migrate non describable entities of type attack-pattern, malware and tool"
+    (let [gen-mutants (mutant-generator [:name :description])
+          source-entities (mapcat #(gen-mutants 5 %) [attack-pattern malware tool])
+          target-entities (map sut/describe source-entities)]
+      (doseq [[source migrated] (zipmap source-entities target-entities)]
+        (let [{src-name :name
+               src-desc :description} source
+
+              {migrated-name :name
+               migrated-title :title
+               migrated-desc :description
+               migrated-short-desc :short_description} migrated]
+
+          (is (nil? migrated-name))
+          (is (seq migrated-title))
+          (is (seq migrated-desc))
+          (is (seq migrated-short-desc))
+
+          (if src-name
+            (is (= src-name migrated-title)
+                "the name key must be renamed into title")
+            (is (= "No title provided" migrated-title)
+                "missing name on corrupted data should be fixed."))
+
+          (when (and (seq src-name) (nil? src-desc))
+            (is (= migrated-title migrated-short-desc)
```

Comment:
`is` inside `when` tends to need extra care as there's the possibility of the assertion not being ever run

For addressing that, you can extract `(and (seq src-name) (nil? src-desc))` as a reusable predicate.

Then, before the doseq over `(zipmap ...)`, you `(assert (some the-predicate (zipmap ...)))`

Reply:
[what about that?](https://github.com/threatgrid/ctia/pull/1065/commits/0862d62bc9b9d6c58a4066a3d30e4ca5317b848f)

I replaced the random generation with a deterministic generation of all mutations in order to avoid ending in a flaky test with the probability (even low) to have a missing mutant and thus a failing `assert`.

Reply:
Bravo 🙌

TIL about `subsets`!

Checking out a sample run:

```clj
> (clojure.math.combinatorics/subsets [1 2 3])
(() (1) (2) (3) (1 2) (1 3) (2 3) (1 2 3))
```

It makes me wonder if ordering matters? e.g one might want to exercise both `x y z` and `z y x`

Reply:
here it's subsets of fields that will be dissoced so the order does not matter, but yes it should be kept in mind for some fonctions.